### PR TITLE
fix: improve stability of Newton Raphson by symmetrizing the Hessian

### PR DESCRIPTION
**Pull request by mandreux-owkin**
## Summary

This PR ensures that the Hessian computed locally in the NewtonRaphson strategy is *numerically* symmetric by computing

```python
hessian = 0.5 * hessian + 0.5 * hessian.T
```

Indeed, floating point errors may make it slightly asymmetric as it is computed by pytorch, resulting in very low eigenvalues. This can be verified by running the introduced test on `main`, which fails without the patch;

## Notes

The bug was originally spotted by @jeandut ;)

## Please check if the PR fulfills these requirements

- [x] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 1]
```diff
@@ -109,6 +109,55 @@ def test_train_newton_raphson_shared_states_results(
     assert pytest.approx(np.array(expected_hessian), rel) == shared_states.hessian
 
 
+@pytest.mark.slow  # the test takes ~30s
+def test_hessian_logistic_regression_newton_raphson(torch_algo, seed):
+    """Test that the Hessian matrix is well computed for a logistic regression problem."""
+    np.random.seed(seed)
+
+    n_samples = 300
+    ndim = 100
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:118]:
```suggestion
def test_hessian_logistic_regression_newton_raphson(torch_algo, seed):
    """Test that the Hessian matrix is well computed for a logistic regression problem."""
    np.random.seed(seed)

    n_samples = 30
    ndim = 10
```

Wdyt about setting the dim to 10 and n sample to 30 ? the test is much faster and also fail without the fix.

**Review by ThibaultFy — COMMENTED** [review]
Thanks a lot for catching and correcting this with a test 😍

File: tests/algorithms/pytorch/test_newton_raphson.py [hunk 2]
```diff
@@ -109,6 +109,55 @@ def test_train_newton_raphson_shared_states_results(
...
+    y_train = np.random.randint(0, 2, n_samples).reshape(-1, 1)
+
+    class LogisticRegression(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = torch.nn.Linear(ndim, 1)
+
+        def forward(self, x):
+            out = self.linear(x)
+            return out
```
Comment (ThibaultFy) [inline tests/algorithms/pytorch/test_newton_raphson.py:129]:
```suggestion
def test_hessian_logistic_regression_newton_raphson(perceptron, torch_algo, seed):
    """Test that the Hessian matrix is well computed for a logistic regression problem."""
    np.random.seed(seed)
    n_samples = 30
    ndim = 10
    
    x_train = np.random.randn(n_samples, ndim)
    y_train = np.random.randint(0, 2, n_samples).reshape(-1, 1)

    model = perceptron(ndim, 1)
```

Comment (ThibaultFy) [reply tests/algorithms/pytorch/test_newton_raphson.py:129]: The suggestion is not well parametrize, sorry, but hope you get the idea to reuse the `perceptron` fixture defined on top of the file :)

File: tests/algorithms/pytorch/test_newton_raphson.py (hunk 1 above)
Comment (mandreux-owkin) [reply tests/algorithms/pytorch/test_newton_raphson.py]: ah thanks for checking! I just reproduced @jeandut 's bug that arose in this dimension and did not take the time to reduce it. It's better if it works in dim 30 indeed

**Comment by mandreux-owkin** [comment]
> Thanks a lot for catching and correcting this with a test 😍

All credits to @jeandut for encountering the error

**Comment by mandreux-owkin** [comment]
>  If its ok for you, we can merge my commit and then your PR to include it in the next week release :)

This is very helpful indeed, thanks! This is merged.

**Review by SdgJlbl — APPROVED** [review]
Thanks for the PR 👍
